aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -726,7 +726,13 @@ Tensor& add_out_dense_sparse_cpu(Tensor& r, const Tensor& dense, const SparseTen
 
 Tensor mul_sparse(const Tensor& self, const Tensor& other) {
   auto commonDtype = at::result_type(self, other);
-  Tensor result = at::empty({0}, self.options().dtype(commonDtype));
+  Tensor result;
+  if (!self.is_sparse()) {
+    result = at::empty({0}, other.options().dtype(commonDtype));
```

Comment:
What about device and such options?

Reply:
The device is stored in `other.options()` and we don't want to use some other device other than `other` tensor lives on. Using tensors on different devices is not allowed here.

Reply:
@mattip - could we add a code comment as to why this is pattern is necessary? I think it's not entirely obvious for an uninitiated reader. Also if you want to avoid the copy constructor (assuming the compiler doesn't take care of it) for result you could write

```
if (self.is_sparse()) {
  return at::mul_out(at::empty({0}, self.options().dtype(commonDtype)), self, other);
}
return at::mul_out(at::empty({0}, other.options().dtype(commonDtype)), self, other);
```

Reply:
The problem is that for any order of arguments we need `result` to be a tensor with strided layout (dense tensor):
_sparse_ self, _dense_ other -> _dense_ result
_dense_ self, _sparse_ other -> _dense_ result

I agree that a comment and `result_options` would be helpful for a future reader.
Only  `result_options` should be different
```
auto result_options = self.is_sparse() ? other.options().dtype(commonDtype) : self.options().dtype(commonDtype);
```

Reply:
commented, adopted change, and rebased